File: cpp/chess/include/pieces.cpp

#include "pieces.h"

inline bool Chess::isPiece(char p) {
	switch (p) { case PAWN: case KNIGHT: case BISHOP: case ROOK: case QUEEN: case KING: return true; }
	return false;
}

inline bool Chess::isCastleLong(std::string move) {
	if (move.length() < 5)
		return false;
	move = move.substr(0,5);
	return (move == "o-o-o" || move == "O-O-O");
}

inline bool Chess::isCastleShort(std::string move) {
	if (move.length() < 3)
		return false;
	move = move.substr(0,3);
	return (move == "o-o" || move == "O-O");
}

inline bool Chess::White::isPiece(char p) {
	switch (p) { case PAWN: case KNIGHT: case BISHOP: case ROOK: case QUEEN: case KING: return true; }
	return false;
}
inline char Chess::White::piece(char p) { return p; }
inline char Chess::White::whichPiece(char p) { return p; }

inline bool Chess::Black::isPiece(char p) {
	switch (p) { case PAWN: case KNIGHT: case BISHOP: case ROOK: case QUEEN: case KING: return true; }
	return false;
}
inline char Chess::Black::piece(char p) { return p + SHIFT; }
inline char Chess::Black::whichPiece(char p) { return p - SHIFT; }

File: src/hs/Chess/Pieces.hs

module Chess.Pieces where

data Piece = Pawn | Knight | Bishop | Rook | Queen | King
    deriving (Enum, Eq, Ord)

pieces :: [Piece]
pieces = enumFrom . toEnum $ 0

piece :: Char -> Maybe Piece
piece 'K' = Just King
piece 'Q' = Just Queen
piece 'R' = Just Rook
piece 'B' = Just Bishop
piece 'N' = Just Knight
piece 'P' = Just Pawn
piece _ = Nothing

instance Show Piece where
    show King = "K"
    show Queen = "Q"
    show Rook = "R"
    show Bishop = "B"
    show Knight = "N"
    show Pawn = "P"

title :: Piece -> Char
title = head . show

value :: Piece -> Int
value Pawn = 1
value Knight = 3
value Bishop = 3
value Rook = 5
value Queen = 9
value King = maxBound


